src/on_chain/on_chain_surplus.py

```diff
@@ -0,0 +1,207 @@
+
+class OnChainEBBO:
+
+    def __init__(self):
+        """
+        Makes ethereum node connection to Infura, and creates contract instance
+        """
+        self.web_3 = Web3(Web3.HTTPProvider(f"https://mainnet.infura.io/v3/{INFURA_KEY}"))
+        self.contract_instance = self.web_3.eth.contract(address=address, abi=gpv2Abi)
+
+
+    def get_hashes_by_block(self, start_block, end_block):
+        """
+        Function filters hashes by contract address
+        """
+        filter_criteria = {
+            'fromBlock': int(start_block),
+            'toBlock': int(end_block),
+            'address': address,
+        }
+        transactions = self.web_3.eth.filter(filter_criteria).get_all_entries()
+        settlement_hashes_list = []
+        for tx in transactions:
+            tx_hash = (tx['transactionHash']).hex()
+            if tx_hash not in settlement_hashes_list:
+                settlement_hashes_list.append(tx_hash)
+        settlement_hashes_list.reverse()
+        return settlement_hashes_list
+
+
+    def get_tx_hashes(self, start_block = None, end_block = None, settlement_hash = None):
+        """
+        Puts transactions in list, calls required function in case of range of blocks.
+        """
+        settlement_hashes_list = []
+        if settlement_hash is not None:
+            settlement_hashes_list = [settlement_hash]
+        elif start_block is not None and end_block is not None:
+            settlement_hashes_list = self.get_hashes_by_block(start_block, end_block)
+            # At this point we have all the needed hashes
+        self.decoder(settlement_hashes_list)
+
+
+    def get_order_data_by_hash(self, settlement_hash):
+        """
+        Returns competition endpoint data since we need order_id and auction_id, and also AWS
+        Bucket response.
+        """
+        try:
+            comp_data = requests.get(f"https://api.cow.fi/mainnet/api/v1/solver_competition/by_tx_hash/{settlement_hash}")
+            status_code = comp_data.status_code
+            if status_code == 200:
+                comp_data = comp_data.json()
+                auction_id = comp_data["auctionId"] 
+                bucket_response = dict(requests.get(f"https://solver-instances.s3.eu-central-1.amazonaws.com/prod-mainnet/{auction_id}.json").json())
+            elif status_code == 404:
+                comp_data = requests.get(f"https://barn.api.cow.fi/mainnet/api/v1/solver_competition/by_tx_hash/{settlement_hash}")
+                status_code = comp_data.status_code
+                if comp_data.status_code == 200:
+                    comp_data = comp_data.json()
+                    auction_id = comp_data["auctionId"] 
+                    bucket_response = dict(requests.get(f"https://solver-instances.s3.eu-central-1.amazonaws.com/staging-mainnet/{auction_id}.json").json())
+        except ValueError as except_err:
+            print("Order not found... " + str(except_err))
+
+        return comp_data, bucket_response
```

Comment:
Although not really a problem, I guess it is a bit wasteful to return all the `comp_data`, since we are only interested in the winning solution.

Reply:
Yeah we can return only the "orders" executed by winning solution, and use them to get id's for matching to bucket_response. That could be an option.